src/gpio.rs

```diff
@@ -23,195 +25,235 @@ mod subscribe_nr {
+            GpioState::Low => false,
+            GpioState::High => true,
         }
     }
 }
 
-impl IrqMode {
-    fn to_num(&self) -> usize {
-        match self {
-            IrqMode::EitherEdge => 0,
-            IrqMode::RisingEdge => 1,
-            IrqMode::FallingEdge => 2,
+impl From<bool> for GpioState {
+    fn from(from_value: bool) -> Self {
+        if from_value {
+            GpioState::Low
+        } else {
+            GpioState::High
         }
     }
 }
 
-pub struct GpioPinUnitialized<'a> {
-    number: usize,
-    phantom: PhantomData<&'a mut ()>,
+pub struct Gpios<'a> {
+    num_gpios: usize,
+    curr_gpio: usize,
+    lifetime: PhantomData<&'a ()>,
 }
 
-pub struct GpioPinWrite<'a> {
-    number: usize,
-    phantom: PhantomData<&'a mut ()>,
-}
+impl<'a> Iterator for Gpios<'a> {
+    type Item = Gpio<'a>;
 
-pub struct GpioPinRead<'a> {
-    number: usize,
-    phantom: PhantomData<&'a mut ()>,
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.curr_gpio < self.num_gpios {
+            let item = Gpio {
+                gpio_num: self.curr_gpio,
+                lifetime: PhantomData,
+            };
+            self.curr_gpio += 1;
+            Some(item)
+        } else {
+            None
+        }
+    }
 }
 
-impl<'a> GpioPinUnitialized<'a> {
-    pub fn open_for_write(self) -> TockResult<GpioPinWrite<'a>> {
-        syscalls::command(DRIVER_NUMBER, command_nr::ENABLE_OUTPUT, self.number, 0)?;
-        Ok(GpioPinWrite {
-            number: self.number,
-            phantom: PhantomData,
-        })
-    }
+pub struct Gpio<'a> {
+    gpio_num: usize,
+    lifetime: PhantomData<&'a ()>,
+}
 
-    pub fn open_for_read(
-        self,
-        callback: Option<(extern "C" fn(usize, usize, usize, usize), IrqMode)>,
-        input_mode: InputMode,
-    ) -> TockResult<GpioPinRead<'a>> {
-        let (callback, irq_mode) = callback.unwrap_or((noop_callback, IrqMode::EitherEdge));
-        self.enable_input(input_mode)
-            .and_then(|pin| pin.subscribe_callback(callback))
-            .and_then(move |pin| pin.enable_callback(irq_mode))
+impl<'a> Gpio<'a> {
+    pub fn enable_output(&mut self) -> TockResult<GpioWrite> {
+        syscalls::command(DRIVER_NUMBER, command_nr::ENABLE_OUTPUT, self.gpio_num, 0)?;
+        let gpio_write = GpioWrite {
+            gpio_num: self.gpio_num,
+            lifetime: PhantomData,
+        };
+        Ok(gpio_write)
     }
 
-    fn subscribe_callback(
-        self,
-        callback: extern "C" fn(usize, usize, usize, usize),
-    ) -> TockResult<GpioPinUnitialized<'a>> {
-        syscalls::subscribe_fn(
-            DRIVER_NUMBER,
-            subscribe_nr::SUBSCRIBE_CALLBACK,
-            callback,
-            self.number,
-        )?;
-        Ok(self)
-    }
-
-    fn enable_input(self, mode: InputMode) -> TockResult<GpioPinUnitialized<'a>> {
+    pub fn enable_input(&mut self, resistor_mode: ResistorMode) -> TockResult<GpioRead> {
```

Comment:
I think change this actually breaks the gpio. I think you have to at least enable the interrupt in order to make reading work. Can you double check this (i.e. using the integration tests)?

Reply:
The integration tests are green.